core/src/packets/ethernet.rs

```diff
@@ -64,18 +48,176 @@ impl fmt::Display for EtherType {
     }
 }
 
-/// Ethernet header.
+/// VLAN tag.
+#[derive(Clone, Copy, Debug, Default)]
+#[repr(C, packed)]
+pub struct VlanTag {
+    tpid: u16,
+    tci: u16,
+}
+
+#[allow(clippy::trivially_copy_pass_by_ref)]
+impl VlanTag {
+    pub fn tag_id(&self) -> u16 {
+        self.tpid
+    }
+
+    pub fn priority(&self) -> u8 {
+        (self.tci >> 13) as u8
+    }
+
+    pub fn drop_eligible(&self) -> bool {
+        self.tci & 0x1000 > 0
+    }
+
+    pub fn identifier(&self) -> u16 {
+        self.tci & 0x0fff
+    }
+}
+
 #[derive(Clone, Copy, Debug, Default)]
 #[repr(C, packed)]
+pub struct Chunk802_1q {
+    tag: VlanTag,
+    ether_type: u16,
+}
+
+#[derive(Clone, Copy, Debug, Default)]
+#[repr(C, packed)]
+pub struct Chunk802_1ad {
+    stag: VlanTag,
+    ctag: VlanTag,
+    ether_type: u16,
+}
+
+/// The ethernet header chunk follows the source mac addr.
+#[derive(Clone, Copy)]
+#[repr(C, packed)]
+pub union Chunk {
+    ether_type: u16,
+    chunk_802_1q: Chunk802_1q,
+    chunk_802_1ad: Chunk802_1ad,
+}
+
+/// Ethernet header.
+#[derive(Clone, Copy)]
+#[repr(C, packed)]
 pub struct EthernetHeader {
     dst: MacAddr,
     src: MacAddr,
-    ether_type: u16,
+    chunk: Chunk,
 }
 
 impl Header for EthernetHeader {}
 
-/// Ethernet packet.
+impl Default for EthernetHeader {
+    fn default() -> EthernetHeader {
+        EthernetHeader {
+            dst: MacAddr::UNSPECIFIED,
+            src: MacAddr::UNSPECIFIED,
+            chunk: Chunk {
+                ether_type: Default::default(),
+            },
+        }
+    }
+}
+
+impl SizeOf for EthernetHeader {
+    /// Size of the ethernet header.
+    ///
+    /// Because the ethernet header is not fixed and modeled with a union, the
+    /// memory layout size is not the correct header size. For a brand new
+    /// ethernet header, we will always report 14 bytes as the fixed portion,
+    /// which is the minimum size without any tags. `Ethernet::header_len()`
+    /// will report the correct instance size based on the presence or absence
+    /// of VLAN tags.
+    #[inline]
+    fn size_of() -> usize {
+        14
+    }
+}
+
+// Tag protocol identifiers.
+const VLAN_802_1Q: u16 = 0x8100;
+const VLAN_802_1AD: u16 = 0x88a8;
+
+/// Ethernet II frame.
```

Comment:
I may re-order the struct to force this up (and have it work w/ docs on a non-pub module), but I'll call it super minor.